pkg/yang/types_builtin.go

```diff
@@ -506,32 +541,97 @@ func ParseNumber(s string) (n Number, err error) {
 		return n, errors.New("sign with no value")
 	}
 
+	n.Kind = Positive
+	ns := s
 	switch s[0] {
 	case '+':
-		s = s[1:]
+		ns = s[1:]
 	case '-':
 		n.Kind = Negative
-		s = s[1:]
+		ns = s[1:]
 	}
-	n.Value, err = strconv.ParseUint(s, 0, 64)
-	return n, err
+	n.Value, err = strconv.ParseUint(ns, 0, 64)
+	if err == nil {
+		return n, nil
+	}
+
+	return DecimalValueFromString(s, -1)
+}
+
+// DecimalValueFromString returns a decimal Number representation of inStr.
+// If fracDigRequired is >= 0, the number is represented with fracDigRequired
+// fractional digits, regardless of the precision of numStr, otherwise the
+// precision of numStr is used to set the number of fractional digits.
+// numStr must conform to Section 9.3.4.
+func DecimalValueFromString(numStr string, fracDigRequired int) (n Number, err error) {
+	s := numStr
+	dx := strings.Index(s, ".")
+	fracDig := 0
+	if dx > 0 {
+		fracDig = len(s) - 1 - dx
+		// remove first decimal, if dx > 1, will fail ParseInt below
+		s = s[:dx] + s[dx+1:]
+	}
+	if fracDigRequired < 0 {
+		fracDigRequired = fracDig
+	}
+	if fracDig > fracDigRequired {
+		return n, fmt.Errorf("%s has too much precision, expect <= %d fractional digits", s, fracDigRequired)
+	}
+
+	for i := 0; i < fracDigRequired-fracDig; i++ {
```

Comment:
s += "000000000000000000"[:fracDigRequired-fracDig]

There has to be at least 18 0's there.